### PR TITLE
Updates readme with more details on configuration

### PR DESCRIPTION
**Pull request by uw-rvitorino**
As directly requested in #36 and also suggested in #25, this PR aims at improving the README with:

- updated configurations for the library versions > 1.1.1
- instructions on setting up a Django application within Keycloak 

Beyond the accuracy of the instructions, revisions to the English grammar and the quality of the screenshots are more then welcome.

### CODE REVIEW

**Comment by moritz89** [comment]
The decode token option is optional, and false by default, so it would make sense to reflect this in the readme. The following would be useful.

DECODE_TOKEN enables extracting a user's profile data from the JWT token instead of fetching this data from Keycloak via an API call. This has the effect of reducing the latency, but also requires the profile to be included in the JWT token in the first place. This can be configured in Keycloak's client scopes tab.

**Comment by moritz89** [comment]
Regarding the Graphene middleware, I'm not sure if we should promote its use. The middleware is basically only used for user authentication in the same way that the normal middleware is, but it is called for each node of the GraphQL query. In effect, this reauthenticates the user multiple times, when once using the normal middleware is enough. Therefore, unless there is some use case that I am not aware of, I would deprecate it instead of recommending its use.

**Comment by moritz89** [comment]
Regarding the setup instructions, they look quite similar to my internal ones. I've adding them below as reference:

## Setup

1. Log into [Keycloak](http://localhost:8080) and go to the "Administration Console"
2. Create a [new realm](http://localhost:8080/auth/admin/master/console/#/create/realm) (sidebar --> Hover over master --> Add realm) with the name `my-realm`
3. [Set up](#set-up-core-service-client) the core service client
4. [Set up](#set-up-web-app-client) the web app client
5. [Set up](#set-up-the-superuser) the superuser

The `imports` folder contains all the settings to use keycloak in a dev environment. It is used by the start script of the core service.

### Set Up Core Service Client

1. In the `my-realm` realm, go to clients and [create a new one](http://localhost:8080/auth/admin/master/console/#/create/client/my-realm)
2. Use `core-service` as the client ID and `http://localhost:8000` as the Root URL and save the client
3. Change the `Access Type` to `confidential` and save the changes
4. Enable `Service Accounts Enabled` and save the changes
5. Go to the `Service Account Roles` tab
6. In the `Client Roles` dropdown select `realm-management`
7. In `Available Roles` select `manage-users` and press `Add selected`
8. Additionally, in `Available Roles` select `view-users` and press `Add selected`
9. This should result in 4 `Effective Roles`: `manage-users`, `query-groups`, `query-users` and `view-users`

In the core service Keycloak settings (core/core/settings.py --> KEYCLOAK_CONFIG) use `core-service` as the client ID and the secret from the `Credentials` tab.

### Set Up Web App Client

1. In the `my-realm` realm, go to clients and [create a new one](http://localhost:8080/auth/admin/master/console/#/create/client/my-realm)
2. Use `web-app` as the client ID and `http://localhost:4200` as the Root URL and save the client

### Set Up the Superuser

1. In the `my-realm` realm, create a [new realm role](http://localhost:8080/auth/admin/master/console/#/create/role/my-realm) (sidebar --> Roles --> Add Role)
2. Name the role `admin` and press `Save`
3. Create a new client role for the `core-service` client (sidebar --> Clients --> core-service --> Roles tab --> Add Role)
4. Name the role `admin` and press  `Save`
5. In the `my-realm` realm, go to the [user management page](http://localhost:8080/auth/admin/master/console/#/realms/my-realm/users)
6. Select `Add user` and create a user with the username `admin`
7. In the `admin` editor, go to the `Role Mappings` tab
8. For the `Realm Roles` --> `Available Roles` select `admin` and the press `Add selected`
9. For the `Client Roles` --> `core-service` select `admin` from the `Available Roles` and press `Add selected`
10. In the `admin` editor, go to the `Credentials` tab
11. Enter the Django superuser password in the `Password` and `Password Confirmation` fields, unselect `Temporary` and save the password.

**Comment by diogosilva30** [comment]
@uw-rvitorino I like the docs improvement. A small, but important, note: These steps (and screenshots) are for keycloak v10. With the latest keycloak v19, they are a bit outdated. Maybe we should target more recent versions?

**Comment by uw-rvitorino** [comment]
Absolutely, I'm a recycling kind of guy 😀, can you, Simão or Moritz
provide such screenshots? Feel free to contribute directly to the branch

*Ricardo Vitorino*

Smart Cities R&I Manager
+351 913 850 631 <+351913850631>

ubiwhere.com

Copyright © Ubiwhere. All rights reserved. Content for exclusive use of
their rightful recipients. Dissemination, copy or use of information by
third parties is prohibited. If you are not an authorised recipient and
received this message, please notify the sender and then delete it without
copying the text and attachments nor distributing or disclosing its
contents.

**Comment by moritz89** [comment]
I'm also still on the legacy Keycloak branch, so my screenshots won't be any newer :sweat_smile:

**Review by diogosilva30 — CHANGES_REQUESTED** [review]
I believe that the documentation README needs to be reviewed, providing information on the new settings. As far as Keycloak integration, I can provide screenshots for recent versions (v19), but I believe that this setup would be better in the repository wiki section.

File: README.md [hunk 1]
```diff
@@ -42,6 +42,7 @@ This package should only be used in projects starting from scratch, since it ove
         'REALM_ADMIN_ROLE': '<REALM_ADMIN_ROLE>',
         'EXEMPT_URIS': [],  # URIS to be ignored by the package
         'GRAPHQL_ENDPOINT': 'graphql/'  # Default graphQL endpoint
+        'DECODE_TOKEN': True  # Required since version 1.1.1
```
Comment (moritz89) [inline README.md:45]: - 'DECODE_TOKEN': True # Optional: Whether to decode token or use Keycloak API to introspect it
- 'VERIFY_AUDIENCE': True # Optional: Whether the token's audience should be verified. Requires Keycloak setup
- 'USER_INFO_IN_TOKEN': True # Optional: If the user info is in the token (used if DECODE_TOKEN is true)
- TRACE_DEBUG_LOGS: False # Optional: Whether to show the traceback of debug logs

File: README.md [hunk 2]
```diff
@@ -42,6 +42,7 @@ This package should only be used in projects starting from scratch, since it ove
         'REALM_ADMIN_ROLE': '<REALM_ADMIN_ROLE>',
         'EXEMPT_URIS': [],  # URIS to be ignored by the package
         'GRAPHQL_ENDPOINT': 'graphql/'  # Default graphQL endpoint
```
Comment (moritz89) [inline README.md:44]: Removed GRAPHQL_ENDPOINT from settings

File: README.md [hunk 3]
```diff
@@ -42,6 +42,7 @@ This package should only be used in projects starting from scratch, since it ove
         'REALM_ADMIN_ROLE': '<REALM_ADMIN_ROLE>',
         'EXEMPT_URIS': [],  # URIS to be ignored by the package
```
Comment (moritz89) [inline README.md:43]: Regex formatted URIs to be ignored by the middleware

**Review by moritz89 — CHANGES_REQUESTED** [review]
Just updates to the Keycloak config section. Rest looks good to me

**Comment by moritz89** [comment]
Hey @uw-rvitorino, how about we just add the 3 small changes I posted above to have the updates visible and do another round of doc improvements in a different MR?

**Pull request closed without merging**